File: src/main/java/org/gids/robot/speech/ASRControlLoop.java
package org.gids.robot.speech;

import edu.cmu.sphinx.api.SpeechResult;
import org.gids.robot.microphone.CustomLiveSpeechRecognizer;

import javax.sound.sampled.AudioInputStream;
import javax.sound.sampled.AudioSystem;
import javax.sound.sampled.Clip;
import java.util.logging.Logger;

/**
 * Created by breandan on 10/23/2015.
 */
public class ASRControlLoop implements Runnable {
    private CustomLiveSpeechRecognizer recognizer;

    public ASRControlLoop(CustomLiveSpeechRecognizer recognizer) {
        // Start-up recognition facilities
        this.recognizer = recognizer;
    }


    private static final Logger logger = Logger.getLogger(ASRControlLoop.class.getSimpleName());

    @Override
    public void run() {
        while (!ListeningState.isTerminated()) {
            // This blocks on a recognition result
            String result = getResultFromRecognizer();
            new SpeechCallback(result);
        }
    }

    private String getResultFromRecognizer() {
        SpeechResult result = recognizer.getResult();

        System.out.println("Recognized: ");
        System.out.println("\tTop H:       " + result.getResult() + " / " + result.getResult().getBestToken() + " / " + result.getResult().getBestPronunciationResult());
        System.out.println("\tTop 3H:      " + result.getNbest(3));

        logger.info("Recognized:    ");
        logger.info("\tTop H:       " + result.getResult() + " / " + result.getResult().getBestToken() + " / " + result.getResult().getBestPronunciationResult());
        logger.info("\tTop 3H:      " + result.getNbest(3));

        return result.getHypothesis();
    }


    // Helpers

    public static synchronized void beep() {
        Thread t = new Thread(() -> {
        try {
            Clip clip = AudioSystem.getClip();
            AudioInputStream inputStream = AudioSystem.getAudioInputStream(
                    ASRService.class.getResourceAsStream("/com.jetbrains.idear/sounds/beep.wav"));
            clip.open(inputStream);
            clip.start();
        } catch (Exception e) {
            System.err.println(e.getMessage());
        }
    });

        t.start();

        try {
            t.join();
        } catch (InterruptedException ignored) {
        }
    }

    private static String splitCamelCase(String s) {
        return s.replaceAll(
                String.format("%s|%s|%s",
                        "(?<=[A-Z])(?=[A-Z][a-z])",
                        "(?<=[^A-Z])(?=[A-Z])",
                        "(?<=[A-Za-z])(?=[^A-Za-z])"
                ),
                " "
        );
    }
}

File: src/main/kotlin/org/gids/robot/speech/ASRControlLoop.kt
package org.gids.robot.speech

import org.gids.robot.microphone.CustomLiveSpeechRecognizer
import org.gids.robot.tts.TTSService
import java.awt.event.KeyEvent.*
import java.io.IOException
import java.util.logging.Level
import java.util.regex.Pattern
import javax.sound.sampled.AudioSystem

class ASRControlLoop(private val recognizer: CustomLiveSpeechRecognizer) : Runnable {
    override fun run() {
        while (!ListeningState.isTerminated) {
            // This blocks on a recognition result
            val result = resultFromRecognizer
            SpeechCallback(result)
        }
    }

    private val resultFromRecognizer: String
        get() {
            val result = recognizer.result

            println("Recognized: ")
            println("\tTop H:       " + result.result + " / " + result.result.bestToken + " / " + result.result.bestPronunciationResult)
            println("\tTop 3H:      " + result.getNbest(3))

            return result.hypothesis
        }

    private fun tellJoke() {
        TTSService.say("knock, knock, knock, knock, knock")

        var result: String? = null
        while ("who is there" != result) {
            result = resultFromRecognizer
        }

        TTSService.say("it's me, turtle")

        try {
            Thread.sleep(3000)
        } catch (e: InterruptedException) {
            e.printStackTrace()
        }

        while (!result!!.contains("wait who") && !result.contains("who are you")) {
            result = resultFromRecognizer
        }

        TTSService.say("turtle")
    }

    private fun pauseSpeech() {
        beep()
        var result: String
        while (ListeningState.isActive) {
            result = resultFromRecognizer
            if (result == "speech resume") {
                beep()
                break
            }
        }
    }


    @Synchronized fun beep() {
        val t = Thread {
            try {
                val clip = AudioSystem.getClip()
                val inputStream = AudioSystem.getAudioInputStream(
                        ASRService::class.java.getResourceAsStream("/org/gids/robot/gids/robot/sounds/beep.wav"))
                clip.open(inputStream)
                clip.start()
            } catch (e: Exception) {
                System.err.println(e.message)
            }
        }

        t.start()

        try {
            t.join()
        } catch (ignored: InterruptedException) {
        }
    }
}
